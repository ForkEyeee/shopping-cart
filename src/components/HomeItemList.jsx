import {
  Box,
  Grid,
  Text,
  Heading,
  Spinner,
  Center,
  HStack,
} from "@chakra-ui/react";
import PropTypes from "prop-types";
import Item from "./HomeItem";

const HomeItemList = ({
  handleAddCart,
  handleClearItems,
  itemQuantity,
  loading,
  error,
  data,
}) => {
  if (loading)
    return (
      <Center p={10}>
        <HStack spacing={5}>
          <Spinner
            thickness="4px"
            speed="0.65s"
            emptyColor="gray.200"
            color="blue.500"
            size="xl"
          />
          <Text>Loading...</Text>
        </HStack>
      </Center>
    );
  if (error) return <Text data-testid="error-message">Oops</Text>;
  return (
    <Box id="itemList">
      <Box>
        <Heading p={5} textAlign={"center"}>
          Today&apos;s Stock
        </Heading>
      </Box>
      <Grid
        templateColumns={{
          base: "repeat(1, 1fr)",
          md: "repeat(2, 1fr)",
          lg: "repeat(2, 1fr)",
          xl: "repeat(3, 1fr)",
        }}
        gap={20}
        p={5}
        className="aaw"
      >
        {data &&
          data.map((item) => (
            <Item
              key={item.id}
              itemId={item.id}
              image={item.image}
              desc={item.description}
              title={item.title}
              handleAddCart={handleAddCart}
              handleClearItems={handleClearItems}
              itemQuantity={itemQuantity}
            >
              <Text
                p={5}
                fontWeight={"semibold"}
                fontSize={"md"}
                noOfLines={2}
                minH={"88px"}
              >
                {item.title}
              </Text>
              <Text
                textAlign={"end"}
                color={"green"}
                fontWeight={"bold"}
                fontSize={32}
              >
                ${item.price}
              </Text>
            </Item>
          ))}
      </Grid>
    </Box>
  );
};

HomeItemList.propTypes = {
  handleAddCart: PropTypes.func.isRequired,
  handleClearItems: PropTypes.func.isRequired,
  itemQuantity: PropTypes.func.isRequired,
};

export default HomeItemList;
