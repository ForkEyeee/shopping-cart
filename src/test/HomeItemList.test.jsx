import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import HomeItemList from "../components/HomeItemList";
import useDataFetching from "../hooks/fetchItemData";
import { renderHook, waitFor } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { ChakraProvider } from "@chakra-ui/react";

describe("useDataFetching Hook", () => {
  it("fetches electronic products and matches the snapshot", async () => {
    const { result } = renderHook(() =>
      useDataFetching("https://fakestoreapi.com/products/")
    );
    await waitFor(() => expect(result.current[0]).not.toBeNull());
    expect(result.current[0]).toMatchSnapshot();
  });
});

it("shows an error message when API call fails", async () => {
  render(
    <ChakraProvider>
      <BrowserRouter>
        <HomeItemList error={true} />
      </BrowserRouter>
    </ChakraProvider>
  );

  const errorMessage = screen.getByText("Oops");
  expect(errorMessage).toBeVisible();
});

it("renders images", async () => {
  const { result } = renderHook(() =>
    useDataFetching("https://fakestoreapi.com/products/")
  );

  await waitFor(() => expect(result.current[0]).not.toBeNull());

  render(
    <ChakraProvider>
      <BrowserRouter>
        <HomeItemList data={result.current[0]} />
      </BrowserRouter>
    </ChakraProvider>
  );

  const imgElements = await screen.findAllByRole("img");
  expect(imgElements.length).toBeGreaterThan(0);
});

it("renders ItemsList Component", async () => {
  render(
    <ChakraProvider>
      <BrowserRouter>
        <HomeItemList />
      </BrowserRouter>
    </ChakraProvider>
  );
  const featuredItemsTitle = await screen.findByText("Today's Stock");
  expect(featuredItemsTitle).toBeVisible();
});
