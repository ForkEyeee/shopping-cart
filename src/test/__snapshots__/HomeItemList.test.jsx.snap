// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useDataFetching Hook > fetches electronic products and matches the snapshot 1`] = `
[
  {
    "description": "Description for Electronic Product 1",
    "id": 1,
    "image": "/assets/images/mock-image.jpg",
    "price": 299.99,
    "title": "Electronic Product 1",
  },
  {
    "description": "Description for Electronic Product 2",
    "id": 2,
    "image": "/assets/images/mock-image.jpg",
    "price": 399.99,
    "title": "Electronic Product 2",
  },
  {
    "description": "Description for Electronic Product 3",
    "id": 3,
    "image": "/assets/images/mock-image.jpg",
    "price": 499.99,
    "title": "Electronic Product 3",
  },
  {
    "description": "Description for Electronic Product 4",
    "id": 4,
    "image": "/assets/images/mock-image.jpg",
    "price": 599.99,
    "title": "Electronic Product 4",
  },
  {
    "description": "Description for Electronic Product 5",
    "id": 5,
    "image": "/assets/images/mock-image.jpg",
    "price": 699.99,
    "title": "Electronic Product 5",
  },
  {
    "description": "Description for Electronic Product 6",
    "id": 6,
    "image": "/assets/images/mock-image.jpg",
    "price": 799.99,
    "title": "Electronic Product 6",
  },
]
`;
